### ISSUE TITLE
Workspaces support

### ISSUE THREAD

**Issue by GMaiolo** [RSeidelsohn/license-checker-rseidelsohn#36]
Been trying to get this to work with a monorepo with workspaces and it's impossible to use the `--production` flag as it yields empty results.

In such environments, each workspace has its own `package.json` but the `node_modules` is located on the root and shared between these workspaces.

Does anyone know if a workaround is possible with this setup?

**Comment by codepunkt** [comment]
What's the point of the `--production` flag anyways? Do you realize that this usually gives you a lot of false positive licenses that are not included in your build output and thus not distributed? Do you understand why this also might give you a few false negatives, missing the licenses of packages that you are distributing?

**Comment by GMaiolo** [comment]
> What's the point of the `--production` flag anyways? Do you realize that this usually gives you a lot of false positive licenses that are not included in your build output and thus not distributed? Do you understand why this also might give you a few false negatives, missing the licenses of packages that you are distributing?

@codepunkt Are you asking what the production flag's point is in my specific scenario or in the library as a whole?
I don't remember what I was trying out at that time (it's been a few months), but I assume I wanted to ignore all internal tooling security issues.

Disregarding the production flag discussion, workspace checks still yield incomplete results.

**Comment by codepunkt** [comment]
@GMaiolo Are the packages/workspaces in your monorepo being built with webpack, rollup or other bundlers?

**Comment by restfulhead** [comment]
Hi. I'm facing the same issue. As a workaround I'm running the check now from the root of our monorepo. But of course that then reports on all dependencies. It would be nicer to be able to run this within each NPM workspace.

@codepunkt We don't use any bundlers, just plain [NPM workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces)

**Comment by RSeidelsohn** [comment]
Hi @restfulhead , @fredrikaverpil just pointed me to this issue - sorry for the delay; this is not a well-maintained repo, as I rarely find enough time to care for it, but it's not abandoned as well.

I never heard of _npm workspaces_ before reading about it here and yet first have to read and understand the documentation. It looks like an important feature, supporting _npm workspaces_, as probably more and more people will use them.

So I'd love to hear from you:

How would you expect or want _license-checker-rseidelsohn_  (sorry for the clumsy name) to behave when it is run on _npm workspaces_? Should it run on all workspaces and separate the output per workspace?

Cheers and thank you for your input, Roman.

**Comment by fredrikaverpil** [comment]
I'll let others chime in as I'm not a ts/node expert, but I was hoping to implement a license checker in our company and we make use of npm workspaces for repos that hosts multiple packages.

- This seems to work _okay_ when sitting at the `$cwd` is at the root `package.json` and `package-lock.json`. I'm hoping that sometime in the future be able to search for all `package.json` files in a repo, and then `cd` into that folder and run this license check tool there. But that does not work currently, when a `package.json` is part of an npm workspace.
- When using npm workspaces, I've noticed that the root `package.json` (which does not have a `license` key by default) acts up; mentioned in #63 .

**Comment by restfulhead** [comment]
Thanks for considering this issue. 

Basically I'd like to be able to run the checker inside each workspace and then it would report the licenses of all dependencies just for that workspace.

For example if I have the following package.json inside a workspace:

```
"scripts": {
    "license-checker": "npx license-checker-rseidelsohn"
}
"devDependencies": {
    "license-checker-rseidelsohn": "4.1.1"
}
"
```

Then I'd like to be able to run `npm run license-checker -w my-workspace` and have it report only relevant dependencies. 

It currently seems to run, but not reporting the correct licenses. Keep in mind that in a workspace setup, there is no `packate-lock.json` file or `node_modules` directory in each workspace; only at the root.

**Comment by RSeidelsohn** [comment]
Hello @restfulhead, @fredrikaverpil, @codepunkt and @GMaiolo,

I will most likely start working on this feature in two weeks (my new employer allows me to work on my OS project every 2nd friday). I finally had a look at the docs and played around with the new `workspaces` feature and think I understand it good enough in order to add a feature for it. But it of course might be that this will become complicated once I start doing it, so I can't promise what the feature will look like exactly or when it might be ready.

Cheers, Roman.

**Comment by RSeidelsohn** [comment]
Hello folks, thanks to @AgentOren, who provided a bugfix PR to the `read-installed-packages` module and then merged the new version 2.0.1 into this module, this issue should now be fixed finally! Awesome.
Also thanks to @Semigradsky, who merged Oren's PR and provides the new 2.0.1 version.

**Comment by huschu** [comment]
This would be a great addition! Are there any updates on this?

**Comment by Schneider1** [comment]
For me it looks as if `read-installed-packages` would need better support for workspaces first. It seems that [this package always looks for the `package.json` and `node-modules` in the same folder](https://github.com/Semigradsky/read-installed/blob/master/read-installed.js#L164). For a workspace however, it would be `./my/workspace/package.json` deeper down and `./node-modules` on top level.

if you run the licence checker in the workspace, it would find the `./my/workspace/package-json`, but no modules; If you run it on top level, it would actually find all your dependencies, but report all which are not in your top level `./package.json` as `extraneous`. This means in turn, that if you use `--production`, only the packages in your top-level `./package.json` are scanned.

**Comment by seantimm** [comment]
It can be more complicated than that as you can have workspaces that end up with a node_modules folder as well as the root path with one (depending on version dependencies).

**Comment by ktarmyshov** [comment]
+1 if there is an update at some point.

But for now as far as I understand, I can use a workaround
clean all node_modules in all workspaces

```
ROOT> clean all node_modules everwhere
ROOT> npm install/ci --workspace=my/workspace
ROOT> npm check:licenses
```

Actually I'm running in a CI this, there it's more important and seems to deliver what I need per workspace, locally do this once in a while.

Cheers.

**Comment by RSeidelsohn** [comment]
Hi folks, sorry for the sloooooooooooow progress with this module (and feel free to participate!). I _still_ struggle a lot with really understanding how the whole code works and why and where. This is (partly) why I am very reluctant taking care of the repo, but I recently came back and am currently continuing my refactoring with the aim of making the code easier to understand. There are still functions of >400 lines of code, and the tests are out of my scope as they use `istanbul` and have a `nyc` configuration that are alien to me (be aware that I took over the repo by cloning it, as the original author of liecense-checker doesn't maintain it any more in every way, even not reacting to my emails - just like I do sometimes, to be fair).
The tests themselves look pretty much like what I'm used to from Jest, but require a better understanding of what the code does how, where and why.

So bear with me, the repo/module isn't dead, but I really struggle making progress. Yet, it's a growth task for me, which is good.

**Comment by ktarmyshov** [comment]
Actually after looking into the whole thing with workspaces, I have now strong doubts, that it's gonna be easy to implement support for workspaces. You're are gonna need to resolve/find each dep and its sub-sub-...-deps declared in the workspace's package.json... to me that seems a lot of work, with very small benefits.



Just a copilot-small-talk


-----

In an npm with workspaces, does root package-lock.json contain any information about which of the installed modules (direct or indirect dependencies) belong to which workspace?
-----

Yes, the root `package-lock.json` in an npm workspace setup does contain information about installed modules, but it does not explicitly indicate which dependencies belong to which workspace. Instead, it maintains a global view of all dependencies across the workspaces, ensuring consistency in package versions.

Each workspace's dependencies are listed under the `dependencies` or `packages` section, but they are not directly mapped to specific workspaces. If you need a dedicated lock file for individual workspaces, there are discussions about generating separate lock files [A](https://github.com/npm/feedback/discussions/650?copilot_analytics_metadata=eyJldmVudEluZm9fY2xpY2tEZXN0aW5hdGlvbiI6Imh0dHBzOlwvXC9naXRodWIuY29tXC9ucG1cL2ZlZWRiYWNrXC9kaXNjdXNzaW9uc1wvNjUwIiwiZXZlbnRJbmZvX2NsaWNrU291cmNlIjoiY2l0YXRpb25MaW5rIiwiZXZlbnRJbmZvX21lc3NhZ2VJZCI6InZhM1hKaEtxelNNcnhHaUV6YmphUSIsImV2ZW50SW5mb19jb252ZXJzYXRpb25JZCI6IjJaQUNpS0c3WURvdFVlakVoR2Y1ayJ9&citationMarker=9F742443-6C92-4C44-BF58-8F5A7C53B6F1). Some developers use tools like `lerna` to manage workspace dependencies more granularly [B](https://stackoverflow.com/questions/74906173/package-lock-json-in-npm-workspaces?copilot_analytics_metadata=eyJldmVudEluZm9fbWVzc2FnZUlkIjoidmEzWEpoS3F6U01yeEdpRXpiamFRIiwiZXZlbnRJbmZvX2NvbnZlcnNhdGlvbklkIjoiMlpBQ2lLRzdZRG90VWVqRWhHZjVrIiwiZXZlbnRJbmZvX2NsaWNrRGVzdGluYXRpb24iOiJodHRwczpcL1wvc3RhY2tvdmVyZmxvdy5jb21cL3F1ZXN0aW9uc1wvNzQ5MDYxNzNcL3BhY2thZ2UtbG9jay1qc29uLWluLW5wbS13b3Jrc3BhY2VzIiwiZXZlbnRJbmZvX2NsaWNrU291cmNlIjoiY2l0YXRpb25MaW5rIn0%3D&citationMarker=9F742443-6C92-4C44-BF58-8F5A7C53B6F1).

Would you like help exploring alternative ways to track dependencies per workspace?

**Comment by ktarmyshov** [comment]
Fixed for myself, sharing with the others, no need to implement anything in the checker. Seems to work.

```bash
#!/bin/bash

# First parameter is the path to the workspace
workspace_path=$1
# List all dependencies of the workspace
deps_json=$(npm ls --workspace="$workspace_path" --json --all)
# Read package name for the workspace
package_name=$(jq -r '.name' "$workspace_path/package.json")

# Recursively extract all dependencies and their versions (dependencies...dependencies...dependencies...)
# Final format: package@full_version
all_deps=$(echo "$deps_json" | jq -r --arg pkg "$package_name" '
  def walkdeps(dep):
    dep | to_entries[] 
    | if (.value.version | type == "string") then
        "\(.key)@\(.value.version)"
      else
        "\(.key)"
      end,
      (if .value.dependencies then walkdeps(.value.dependencies) else empty end);
  .dependencies[$pkg].dependencies | walkdeps(.)
' | sort -u)

# echo "$all_deps"

# Replace all whitespace with semicolons and remove the last semicolon
all_deps_check=$(echo "$all_deps" | tr -s '[:space:]' ';' | sed 's/;$//')

# Check the licenses of extracted dependencies
npx license-checker-rseidelsohn --includePackages "$all_deps_check" --json
```

**Comment by fnajera-rac-de** [comment]
After migrating to a monorepo with workspaces we ran into the same problem. This script from @ktarmyshov is great! 

In case it's useful for anyone: we are running it from Windows (not WSL) so I had to convert it from bash, and then we ran into a second problem: `$all_deps_check` is humongous and we hit "command line is too long". I implemented a small quick-and-dirty patch so that I can write that list into a temp file and pass it to the tool:

```diff
diff --git a/node_modules/license-checker-rseidelsohn/lib/indexHelpers.js b/node_modules/license-checker-rseidelsohn/lib/indexHelpers.js
index ba7b25c..4d5ac64 100644
-- a/node_modules/license-checker-rseidelsohn/lib/indexHelpers.js
++ b/node_modules/license-checker-rseidelsohn/lib/indexHelpers.js
@@ -93,6 +93,9 @@ const getOptionArray = (option) => {
     }
 
     if (typeof option === 'string') {
+        if (option.startsWith('file:')) {
+            return fs.readFileSync(option.substring(5), 'utf8').split(';');
+        }
         return option.split(';');
     }
```

Then call `npx license-checker-rseidelsohn --includePackages "file:path-to-temp-file" --json`
